src/policy/lockfreeimmortalspace.rs

```diff
@@ -108,15 +108,20 @@ impl<VM: VMBinding> Space<VM> for LockFreeImmortalSpace<VM> {
     }
 
     fn reserved_pages(&self) -> usize {
-        let cursor = unsafe { Address::from_usize(self.cursor.load(Ordering::Relaxed)) };
+        let cursor = self.cursor.load(Ordering::Relaxed);
         let data_pages = conversions::bytes_to_pages_up(self.limit - cursor);
         let meta_pages = self.metadata.calculate_reserved_pages(data_pages);
         data_pages + meta_pages
     }
 
     fn acquire(&self, _tls: VMThread, pages: usize) -> Address {
         let bytes = conversions::pages_to_bytes(pages);
-        let start = unsafe { Address::from_usize(self.cursor.fetch_add(bytes, Ordering::Relaxed)) };
+        let start = self
+            .cursor
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |addr| {
```

Comment:
It should be fine no matter the closure is inlined or not. It is an atomic operation on the slow path. By the way, why not just use fetch_add so the code should be identical to what it was before?

Reply:
that's the problem of 'atomic' that 'fetch_add' is not implemented for Atomic<Address>, only integer types got supported.

Reply:
There is a different atomic library that might support `fetch_add` if you derive the `AtomicInteger` implementation: https://docs.rs/atomig/latest/atomig/trait.AtomInteger.html. Might be worth looking into

Reply:
> There is a different atomic library that might support `fetch_add` if you derive the `AtomicInteger` implementation: https://docs.rs/atomig/latest/atomig/trait.AtomInteger.html. Might be worth looking into

Do you mean wrapping a fetch_add manually? Unfortunately, `AtomInteger` does not support `fetch_add` with type `T`, which implemented the bound of `Self: AddAssign<T>`. Instead, only `Wrapper<Self>::fetch_add(Self, Ordering) -> Self` is supported. I don't really consider adding `Address` together a good idea, since you don't do so.

To make it possible to fetch_add other types while stay on primitive atomic transactions requires some unsafe and effort. I don't have the confidence to write it out beautifully.